### ISSUE TITLE
References not working

### ISSUE THREAD

**Issue by hiperterminal** [Zettlr/Zettlr#1216]
## Description
After include an "@", Zettlr is not displaying the list of references.

## Reproducing
Export from Zotero your bibliography in different formats, .bib, .json, etc. and then while you are writing, include an "@" to start searching for your references. I try it on Windows and Ubuntu and I guess it's because Zettlr just recognizes a specific .bib file structure and Zotero exported in another structure. none of the .json or .csl files are recognized.

## Expected behaviour

If you type an "@" a list should appear. Nothing happens.

## Platform
 - OS and version: Ubuntu 20.05
 - Zettlr Version: 1.7.5

**Comment by canpolat** [comment]
@hiperterminal I use this functionality quite often and have not experienced any problems on Windows 10, Zettlr version 1.7.5.

I have had some hiccups, but all of them due to missing a necessary aspect of enabling citation. As you probably know, the functionality is documented [here](https://docs.zettlr.com/en/academic/citations/#citing-with-zettlr). Could you please go through the steps and double check if you have skipped one of the steps.

From my own experience with importing citations, usual suspects are:
- You have not installed the BetterBibTex plugin
- You are not exporting the library in correct format
- The file path for the library is not set correctly in Zettlr (Preferences -> Export)

**Comment by hiperterminal** [comment]
> @hiperterminal I use this functionality quite often and have not experienced any problems on Windows 10, Zettlr version 1.7.5.
> 
> I have had some hiccups, but all of them due to missing a necessary aspect of enabling citation. As you probably know, the functionality is documented [here](https://docs.zettlr.com/en/academic/citations/#citing-with-zettlr). Could you please go through the steps and double check if you have skipped one of the steps.
> 
> From my own experience with importing citations, usual suspects are:
> 
>     * You have not installed the BetterBibTex plugin
> 
>     * You are not exporting the library in correct format
> 
>     * The file path for the library is not set correctly in Zettlr (Preferences -> Export)

@canpolat Thanks, I double checked, but is not working. I have a .bib file from December-2019 and it works, but when I export my library from Zotero, in .json or .bib, Zettlr doesn't recognize it. I use this function but whitout adding new references after 2019.

**Comment by canpolat** [comment]
I suppose, if you could share a minimal exported JSON file (maybe with only 1 or 2 titles in it) that does not work with Zettlr, I could try that in my system and see if I can reproduce the problem.

This is how the "Automatic export" tabs looks like in my Zotero:

![image](https://user-images.githubusercontent.com/459891/91729432-d5104a80-eba4-11ea-87e9-249c66f6e286.png)

**Comment by hiperterminal** [comment]
> I suppose, if you could share a minimal exported JSON file (maybe with only 1 or 2 titles in it) that does not work with Zettlr, I could try that in my system and see if I can reproduce the problem.

Hi. I exported as you showed me with some records and the reference works well. The problem is when I try to export all my library [attached here](https://github.com/Zettlr/Zettlr/files/5187609/biblioteca-zotero.zip).

I hope it helps. Thanks.

**Comment by canpolat** [comment]
## Analysis

I downloaded the file and added it to my Zettlr instance as the source for citations. And after a restart, I could reproduce the problem. That was almost a certain sign that the problem was with the JSON file. When I opened the logs I saw the following errors:

![2020-09-08_17-15-32](https://user-images.githubusercontent.com/459891/92495312-0ddc9f00-f1f7-11ea-9e28-05c951cfc9bc.png)

So, without going into details of what the problem may be (possibly the `ö` in the name), I manually removed that reference from the JSON file and restarted Zettlr. After the restart, the problem identifier this time was `Fernández-de-Arroyabe-Olaortua201861` and then `Krötzsch2007251`, then `Santillán2014522`. It looks like these 5 items are causing the problem:

- Döring20171016
- Fernández-de-Arroyabe-Olaortua201861
- Krötzsch2007251
- Santillán2014522
- Tüür-Fröhlich2011279

After removing all these, Zettlr could successfully render the citations:

![image](https://user-images.githubusercontent.com/459891/92498537-0ae3ad80-f1fb-11ea-8f33-df0ec0bd3858.png)

## What now?

First of all, I understand removing citations is not an option (it is not a general enough solution). So, I wanted to go deeper and find the root cause. After a quick lookup, I found that the problem may be this setting for you:

![image](https://user-images.githubusercontent.com/459891/92497218-58f7b180-f1f9-11ea-9cf4-bc4587be8092.png)

[Better BibTeX documentation](https://retorque.re/zotero-better-bibtex/installation/preferences/citation-keys/) mentions this: 

![image](https://user-images.githubusercontent.com/459891/92497311-7c226100-f1f9-11ea-8ed8-e39d942e3856.png)

Could you please check what this is set to in you **Zotero Preferences**?

## Solution
So, here is a plan (that is a bit tedious and has some risks, **backup all your files and the JSON file before starting this**):

0. Close Zettlr
1. Check the checkbox for forcing citation keys to plain text and refresh the citation keys for these 5 items manually (you will have to find and right click on each one of these and refresh the key, see screenshot 1 below).
2. Make sure that the new 'citekey' does not have the non-ASCII characters
3. Launch Zettlr, create a new document and test that auto-complete works. If it does not, then the problem is not solved so do not continue.
4. If the previous step went fine, in Zettlr search for each one of these references and manually update them to their newer versions. For example, you can search for `@Döring20171016` and find all the documents where that specific key is being used. You will now have the help of the auto-complete.

I think this should solve the problem for you.

**Screenshot 1:**
![image](https://user-images.githubusercontent.com/459891/92498916-83e30500-f1fb-11ea-94b5-2db856941bdc.png)

**Comment by canpolat** [comment]
## Moving forward
@nathanlesage, This issue shows that Zettlr cannot handle non-ASCII characters in citation kesy. I did not find this documented this anywhere (although Better BibTeX defaults create valid/ASCII keys). Should this be documented or fixed?

**Comment by nathanlesage** [comment]
Would you mind testing this with the Citr-library directly? Zettlr uses this for everything citation rendering

**Comment by hiperterminal** [comment]
> After removing all these, Zettlr could successfully render the citations

Thanks. Now it works.

**Comment by canpolat** [comment]
> Would you mind testing this with the Citr-library directly? Zettlr uses this for everything citation rendering

If this was meant for me, I don't really have a setup to do this (node, etc.), but Citr documentation seems to give an answer:

```js
const Citr = require('Citr')

let asciiKey = '@Doe1990'
let unicodeKey = '@村上2018'

Citr.util.validateCitationID(asciiKey) // true
Citr.util.validateCitationID(asciiKey, true) // true (strict mode enabled)
Citr.util.validateCitationID(unicodeKey) // true (Japanese characters are allowed)
Citr.util.validateCitationID(unicodeKey, true) // false (only ASCII characters allowed)

try {
  let citation = Citr.parseSingle(unicodeKey, true) // Enable strict mode
} catch (err) {
  console.error('An error will be thrown, as parseSingle will call validateCitationID using strict mode')
}
```

If I'm looking at the correct place, Zettlr does not seem to use 'strict' mode via `parseSingle`, so this may indeed be an issue in Citr.

https://github.com/Zettlr/Zettlr/blob/b1e1e36d962ad1d20a61cca46817208fab487cdd/source/main/providers/citeproc-provider.js#L424

**Comment by nathanlesage** [comment]
> > Would you mind testing this with the Citr-library directly? Zettlr uses this for everything citation rendering
> 
> 
> 
> If this was meant for me, I don't really have a setup to do this (node, etc.), but Citr documentation seems to give an answer:
> 
> 
> 
> ```js
> 
> const Citr = require('Citr')
> 
> 
> 
> let asciiKey = '@Doe1990'
> 
> let unicodeKey = '@村上2018'
> 
> 
> 
> Citr.util.validateCitationID(asciiKey) // true
> 
> Citr.util.validateCitationID(asciiKey, true) // true (strict mode enabled)
> 
> Citr.util.validateCitationID(unicodeKey) // true (Japanese characters are allowed)
> 
> Citr.util.validateCitationID(unicodeKey, true) // false (only ASCII characters allowed)
> 
> 
> 
> try {
> 
>   let citation = Citr.parseSingle(unicodeKey, true) // Enable strict mode
> 
> } catch (err) {
> 
>   console.error('An error will be thrown, as parseSingle will call validateCitationID using strict mode')
> 
> }
> 
> ```
> 
> 
> 
> If I'm looking at the correct place, Zettlr does not seem to use 'strict' mode via `parseSingle`, so this may indeed be an issue in Citr.
> 
> 
> 
> https://github.com/Zettlr/Zettlr/blob/b1e1e36d962ad1d20a61cca46817208fab487cdd/source/main/providers/citeproc-provider.js#L424

Unfortunately not, as it doesn't use strict mode (= not passing `true`) :/

**Comment by canpolat** [comment]
> Unfortunately not, as it doesn't use strict mode (= not passing `true`) :/

These two examples make me think that not using 'strict' makes it allow non-ASCII characters, but you obviously have a better understanding of the code, so I'll let you decide :)

```js
Citr.util.validateCitationID(unicodeKey) // true (Japanese characters are allowed)
Citr.util.validateCitationID(unicodeKey, true) // false (only ASCII characters allowed)
```

**Comment by canpolat** [comment]
Yes, but Zettlr is not passing `true`:

https://github.com/Zettlr/Zettlr/blob/b1e1e36d962ad1d20a61cca46817208fab487cdd/source/main/providers/citeproc-provider.js#L424

Anyway, I think we at least know the root cause of the behavior. Whether it is on Citr or Zettlr is secondary, in my opinion.

**Comment by nathanlesage** [comment]
> After removing all these, Zettlr could successfully render the citations:

Short update for this: I accidentally discovered a big, logical error in the Citr library, which would apply the strict citekey mode in the last version. I have fixed this bug now, and it will ship with 1.8 – then you can again use Umlauts and other non-ASCII-characters just fine. I would like to apologise for this very stupid and preventable error.
